src/v/kafka/server/snc_quota_manager.cc

```diff
@@ -310,8 +315,212 @@ ss::future<> snc_quota_manager::quota_balancer_step() {
+                              const quota_t share = muldiv(
+                                delta, surplus.at(k), total_surplus);
+                              schedule.at(k) += share;
+                              remainder -= share;
+                          }
+                          // the remainder equally
+                          vassert(
+                            remainder <= 0,
+                            "Expected {} <= 0; delta: {}, total_surplus: {}, "
+                            "surpluses: {}",
+                            remainder,
+                            delta,
+                            total_surplus,
+                            surplus);
+                          if (remainder < 0) {
+                              const quota_t d = (remainder + 1)
+                                                  / static_cast<quota_t>(
+                                                    schedule.size())
+                                                - 1;
+                              for (quota_t& s : schedule) {
+                                  const quota_t dd = std::max(d, remainder);
+                                  remainder -= dd;
+                                  s += dd;
+                              }
+                          }
+                      } else { // delta <= -total_surplus
+
+                          // all surpluses
+                          for (size_t k = 0; k != ss::smp::count; ++k) {
+                              const quota_t share = -surplus.at(k);
+                              schedule.at(k) += share;
+                              delta -= share;
+                          }
+                          // the rest equally
+                          auto share = std::div(delta, ss::smp::count);
+                          for (quota_t& s : schedule) {
+                              s += share.quot;
+                              if (share.rem < 0) {
+                                  s += -1;
+                                  share.quot -= -1;
+                              }
+                          }
+                      }
+                  }
+              },
+              schedule,
+              deltas,
+              total_surplus,
+              to_inside_out(surplus));
+        }
+        if (deltas.in > 0 || deltas.out > 0) {
+            to_each(
+              [](std::vector<quota_t>& schedule, quota_t& delta) {
+                  if (delta > 0) {
+                      // equally
+                      auto share = std::div(delta, ss::smp::count);
+                      for (quota_t& s : schedule) {
+                          s += share.quot;
+                          if (share.rem > 0) {
+                              s += 1;
+                              share.quot -= 1;
+                          }
+                      }
+                  }
+              },
+              schedule,
+              deltas);
+        }
```

Comment:
definitely needs explicit comments about what is happening.

Reply:
I think this code is self doc now